File: carFleet.cpp
//stack
//time complexity = O(nlogn)
//space complexity O(n)

#import <vector>
#import <stack>
#import <utility> //for pair keyword
class Solution{
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed){
        std::vector<std::pair<int,int>> pair;
        for(int i = 0; i < position.size(), i++){
            pair.push_back({position[i], speed[i]});
        } //populated pairs for car position and speed

        std::sort(pair.rbegin(), pair.rend()); //r is reverse (sort in reverse order)

        std::stack<double> stack;
        //auto tells compiler to deduce type of variable p
        //& p is a reference to each element of container pair
        for(auto& p : pair){
            double time = ((double)(target - p.first)/ p.second); //target-position = distance to target, distance to target / speed = the time to get there
            if (!stack.empty() && time < stack.top()){
                stack.pop()
            }
            //push the current car's time onto stack
            stack.push(time);
        }
        return stack.size() // return number of fleets
    }
}

File: carFleet.py
#stack
#time complexity -> O(nlogn) due to using the sorted algorithm + forloop
#space complexity -> O(n) memory due to using stack
from typing import List
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        pair = [[p,s] for p,s in zip(position,speed)]
        #create an array of pairs

        stack = []
        for p,s in sorted(pair)[::-1]: #reverse sorted order
            stack.append((target - p) / s)
            if len(stack) > 1 and stack[-1] <= stack[-2]: #if we have 2 cars in stack
            #and car at top of stack is going faster than car in front of it
                stack.pop()
        return len(stack)